perf: GIZMO should not check for unknown classes in Quarkus

Christopher-Chianelli opened the pull request:
In Quarkus, we know all subclasses, so we cannot encounter an unknown class.

Christopher-Chianelli committed:
perf: GIZMO should not check for unknown classes in Quarkus

In Quarkus, we know all subclasses, so we cannot encounter
an unknown class.

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/domain/solution/cloner/gizmo/GizmoSolutionClonerImplementor.java:
```diff
@@ -52,8 +52,11 @@ public class GizmoSolutionClonerImplementor {
+    private static final String FALLBACK_CLONER = "fallbackCloner";
     public static final boolean DEBUG = false;
 
+    public static boolean IS_QUARKUS = false;
```
Let's figure out a different way to pass this information. Static mutable fields are fundamentally wrong, and we've seen it numerous times already.

Do we need a Quarkus-specific subtype? So be it.
Quarkus-specific constructor? Sure.
Just not this, please.

In my opinion, the fact that there are ifs for particular situations screams for two distinct implementations with a shared supertype, or the lambda/strategy pattern.

triceo reviewed (changes requested):
Thanks for the adjustment. Please see the comment and let's figure out a way to do the switch better.
Also, CI is failing.

Christopher-Chianelli committed:
chore: Make QuarkusSolutionClonerImplmentor methods' virtual

This allows the Quarkus cloner to override them without using
a static mutable variable.

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/domain/solution/cloner/gizmo/GizmoSolutionClonerFactory.java:
```diff
@@ -28,7 +28,8 @@ public static <T> SolutionCloner<T> build(SolutionDescriptor<T> solutionDescript
                     "Maybe add a dependency to io.quarkus.gizmo:gizmo.");
         }
-        return GizmoSolutionClonerImplementor.createClonerFor(solutionDescriptor, gizmoClassLoader);
+        return new GizmoSolutionClonerImplementor().createClonerFor(solutionDescriptor,
```
Why not:

```suggestion
        return GizmoSolutionClonerImplementor.createClonerFor(solutionDescriptor,
```

Seems like the instance creation is just boilerplate, it can happen internally. People need not know what's happening under the covers.

Christopher-Chianelli replied:
See comment below

triceo commented on core/core-impl/src/main/java/ai/timefold/solver/core/impl/domain/solution/cloner/gizmo/GizmoSolutionClonerImplementor.java:
```diff
@@ -100,14 +101,20 @@ public static Comparator<Class<?>> getInstanceOfComparator(Set<Class<?>> deepClo
+    protected void createFields(ClassCreator classCreator) {
+        classCreator.getFieldCreator(FALLBACK_CLONER, FieldAccessingSolutionCloner.class)
+                .setModifiers(Modifier.PRIVATE | Modifier.STATIC);
+    }
```
Could have stayed static.

Christopher-Chianelli replied:
No, this method is overridden in Quarkus to do nothing, which make it a mandatory to be virtual.

triceo commented on quarkus-integration/quarkus/deployment/src/main/java/ai/timefold/solver/quarkus/deployment/GizmoMemberAccessorEntityEnhancer.java:
```diff
@@ -417,6 +419,76 @@ public String generateGizmoBeanFactory(ClassOutput classOutput, Set<Class<?>> be
         return generatedClassName;
     }
 
+    private static class QuarkusGizmoSolutionClonerImplementor extends GizmoSolutionClonerImplementor {
```
The class is IMO a big too big to be an inner class. Consider making it a top-level class; it's already static, so you don't need the reference to the outer class.

triceo commented on core/core-impl/src/test/java/ai/timefold/solver/core/impl/domain/solution/cloner/gizmo/GizmoSolutionClonerTest.java:
```diff
@@ -69,7 +69,7 @@ protected <Solution_> SolutionCloner<Solution_> createSolutionCloner(SolutionDes
                 });
 
-        GizmoSolutionClonerImplementor.defineClonerFor(classCreator, solutionDescriptor,
+        new GizmoSolutionClonerImplementor().defineClonerFor(classCreator, solutionDescriptor,
```
Same comment as above; I think it could've been made static and hidden the instance.

Christopher-Chianelli replied:
Issue is, Quarkus need to pass its own subclass as an instance instead of using the default one; I prefer `new Subclass().defineClonerFor(args)` to `Implementor.defineClonerFor(new Subclass(), args)`

triceo replied:
> Issue is, Quarkus need to pass its own subclass as an instance instead of using the default one; I prefer `new Subclass().defineClonerFor(args)` to `Implementor.defineClonerFor(new Subclass(), args)`

In the example you give, I can see how that is a question of personal preference.
But on line 72, where I put my comment, you're creating the instance without assigning it to any variable; why have the user create the instance then? It just looks odd.

triceo reviewed (approved):
Some of these newly instance methods might have been kept static, I think - let's do that.  
Also some more smaller comments.

Once resolved, good to merge.

Christopher-Chianelli committed:
chore: Make GIZMO API methods static with overload for Quarkus
